datafusion/core/src/physical_plan/file_format/row_filter.rs

```diff
@@ -0,0 +1,367 @@
+
+                if !is_primitive_field(self.file_schema.field(idx)) {
+                    self.non_primitive_columns = true;
+                }
+            } else if self.table_schema.index_of(&column.name).is_err() {
+                // If the column does not exist in the (un-projected) table schema then
+                // it must be a projected column.
+                self.projected_columns = true;
+            }
+        }
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::Column(Column { name, .. }) = &expr {
+            if self.file_schema.field_with_name(name).is_err() {
+                return Ok(Expr::Literal(ScalarValue::Null));
+            }
+        }
+
+        Ok(expr)
+    }
+}
+
+/// Calculate the total compressed size of all `Column's required for
+/// predicate `Expr`. This should represent the total amount of file IO
+/// required to evaluate the predicate.
+fn size_of_columns(
+    columns: &HashSet<Column>,
+    schema: &Schema,
+    metadata: &ParquetMetaData,
+) -> Result<usize> {
+    let mut total_size = 0;
+    let row_groups = metadata.row_groups();
+    for Column { name, .. } in columns.iter() {
+        let idx = schema.index_of(name)?;
+
+        for rg in row_groups.iter() {
+            total_size += rg.column(idx).compressed_size() as usize;
+        }
+    }
+
+    Ok(total_size)
+}
+
+/// For a given set of `Column`s required for predicate `Expr` determine whether all
+/// columns are sorted. Sorted columns may be queried more efficiently in the presence of
+/// a PageIndex.
+fn columns_sorted(
+    _columns: &HashSet<Column>,
+    _schema: &Schema,
+    _metadata: &ParquetMetaData,
+) -> Result<bool> {
+    // TODO How do we know this?
+    Ok(false)
+}
+
+/// Build a [`RowFilter`] from the given predicate `Expr`
+pub fn build_row_filter(
+    expr: Expr,
+    file_schema: &Schema,
+    table_schema: &Schema,
+    metadata: &ParquetMetaData,
+) -> Result<Option<RowFilter>> {
+    let predicates = disjoin_filters(expr);
+
+    let candidates: Vec<FilterCandidate> = predicates
+        .into_iter()
+        .flat_map(|expr| {
+            if let Ok(candidate) =
+                FilterCandidateBuilder::new(expr.clone(), file_schema, table_schema)
+                    .build(metadata)
+            {
+                candidate
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    if candidates.is_empty() {
+        return Ok(None);
+    }
+
+    let (indexed_candidates, mut other_candidates): (Vec<_>, Vec<_>) =
+        candidates.into_iter().partition(|c| c.can_use_index);
+
+    let mut filters: Vec<Box<dyn ArrowPredicate>> = vec![];
+
+    for candidate in indexed_candidates {
+        let filter = DatafusionArrowPredicate::try_new(candidate, file_schema, metadata)?;
+
+        filters.push(Box::new(filter));
+    }
+
+    other_candidates.sort_by_key(|c| c.required_bytes);
```

Comment:
Perhaps we could move this ahead of the partition, so that it also used to sort the indexed expressions?